core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeAvailabilityService.java

```diff
@@ -81,24 +77,29 @@
     @Inject
     private ClusterComponentConfigProvider clusterComponentConfigProvider;
 
-    public UpgradeV4Response checkForUpgradesByName(Long workspaceId, String stackName, boolean lockComponents) {
-        Stack stack = stackService.getByNameInWorkspace(stackName, workspaceId);
+    public UpgradeV4Response checkForUpgradesByName(Stack stack, boolean lockComponents, Boolean replaceVms) {
         UpgradeV4Response upgradeOptions = checkForUpgrades(stack, lockComponents);
         if (StringUtils.isEmpty(upgradeOptions.getReason())) {
-            Result<Map<HostGroupName, Set<InstanceMetaData>>, RepairValidation> validationResult = clusterRepairService.repairWithDryRun(stack.getId());
             if (!stack.getStatus().isAvailable()) {
                 upgradeOptions.setReason(String.format("Cannot upgrade cluster because it is in %s state.", stack.getStatus()));
                 LOGGER.warn(upgradeOptions.getReason());
-            } else if (validationResult.isError()) {
-                upgradeOptions.setReason(String.join(",", validationResult.getError().getValidationErrors()));
-                LOGGER.warn(String.format("Cannot upgrade cluster because: %s", upgradeOptions.getReason()));
+            } else if (shouldValidateForRepair(lockComponents, replaceVms)) {
+                LOGGER.debug("Validate for repair");
+                Result<Map<HostGroupName, Set<InstanceMetaData>>, RepairValidation> validationResult = clusterRepairService.repairWithDryRun(stack.getId());
+                if (validationResult.isError()) {
+                    upgradeOptions.setReason(String.join(",", validationResult.getError().getValidationErrors()));
+                    LOGGER.warn(String.format("Cannot upgrade cluster because: %s", upgradeOptions.getReason()));
+                }
             }
         }
         return upgradeOptions;
     }
 
-    public UpgradeV4Response checkForNotAttachedClusters(StackViewV4Responses stackViewV4Responses, UpgradeV4Response upgradeOptions) {
+    private boolean shouldValidateForRepair(boolean lockComponents, Boolean replaceVms) {
+        return lockComponents || replaceVms == null || replaceVms;
```

Comment:
not really, it would return false in case `replaceVms` is null. I could replace with not `!Boolean.FALSE.equals(replaceVms)`, but my guess the intention this way is more visible. Let me know if you still would like me to change it.